src/security/hapi-auth-validation.js

```diff
@@ -1,68 +1,134 @@
 'use strict'
 
 const async = require('async')
+const Boom = require('boom')
 
+const config = require('./../lib/config')
 const authConfig = require('./../lib/config/config.auth')
-
 const userOps = require('./../lib/ops/userOps')
-const authorize = require('./../lib/ops/authorizeOps')
+const authorizeOps = require('./../lib/ops/authorizeOps')
+
+function isSuperUser (user) {
+  return user.organizationId === config.get('authorization.superUser.organization.id')
+}
+
+function loadUser (job, next) {
+  const { userId } = job
+
+  userOps.getUserOrganizationId(userId, (err, organizationId) => {
+    if (err) return next(Boom.unauthorized('Bad credentials'))
+
+    userOps.readUser({ id: userId, organizationId }, (err, user) => {
+      if (err) return next(Boom.unauthorized('Bad credentials'))
+      job.currentUser = user
+
+      next()
+    })
+  })
+}
+
+function impersonate (job, next) {
+  const { currentUser } = job
+  job.organizationId = currentUser.organizationId
+
+  if (isSuperUser(currentUser) && job.requestedOrganizationId) {
+    job.organizationId = job.requestedOrganizationId
+  }
+
+  next()
+}
+
+function checkAuthorization (userId, action, resource, done) {
+  const params = { userId, action, resource }
+
+  authorizeOps.isUserAuthorized(params, (err, result) => {
+    if (err) return done(err)
+    done(null, result.access)
+  })
+}
+
+function buildResourcesForUser (builder, userId, organizationId, done) {
+  const buildParams = {
+    userId,
+    teamId: '*',
+    organizationId
+  }
+
+  const resources = [builder(buildParams)]
+
+  userOps.readUser({ id: userId, organizationId: organizationId }, (err, user) => {
+    if (err && err.output.statusCode === 404) return done(null, resources)
+    if (err) return done(err)
+
+    user.teams.forEach((team) => {
+      buildParams.teamId = team.id
+      resources.push(builder(buildParams))
+    })
+
+    done(null, resources)
+  })
+}
+
+function buildResources (plugin, request, organizationId, done) {
+  let resource = plugin.resource
+
+  if (resource) {
+    return done(null, [resource])
+  }
+
+  const resourceType = request.route.path.split('/')[2]
+  const resourceBuilder = authConfig.resources[resourceType]
+
+  if (!resourceBuilder) {
+    return done({ error: 'Resource builder not found' })
```

Comment:
I don't think so. In all honestly this error will be just wrapped in a boom error and the message lost (leaking the actual reason why authorisation fails can be a security bug).

I kept it here mostly out of laziness, but it may become useful if we want to log the actual error